ml-agents/mlagents/trainers/training_status.py

```diff
@@ -0,0 +1,114 @@
+from typing import Dict, Any
+from enum import Enum
+from collections import defaultdict
+import json
+import attr
+import cattr
+
+from mlagents.tf_utils import tf
+from mlagents_envs.logging_util import get_logger
+from mlagents.trainers import __version__
+from mlagents.trainers.exception import TrainerError
+
+logger = get_logger(__name__)
+
+STATUS_FORMAT_VERSION = "0.1.0"
+
+
+class StatusType(Enum):
+    LESSON_NUM = "lesson_num"
+    STATS_METADATA = "metadata"
+
+
+@attr.s(auto_attribs=True)
+class StatusMetaData:
+    stats_format_version: str = STATUS_FORMAT_VERSION
+    mlagents_version: str = __version__
+    tensorflow_version: str = tf.__version__
+
+    def to_dict(self) -> Dict[str, str]:
+        return cattr.unstructure(self)
+
+    @staticmethod
+    def from_dict(import_dict: Dict[str, str]) -> "StatusMetaData":
+        return cattr.structure(import_dict, StatusMetaData)
+
+    def check_compatibility(self, other: "StatusMetaData") -> None:
+        """
+        Check compatibility with a loaded StatsMetaData and warn the user
+        if versions mismatch. This is used for resuming from old checkpoints.
+        """
+        # This should cover all stats version mismatches as well.
+        if self.mlagents_version != other.mlagents_version:
+            logger.warning(
+                "Checkpoint was loaded from a different version of ML-Agents. Some things may not resume properly."
+            )
+        if self.tensorflow_version != other.tensorflow_version:
+            logger.warning(
+                "Tensorflow checkpoint was saved with a different version of Tensorflow. Model may not resume properly."
+            )
+
+
+class GlobalTrainingStatus:
+    saved_state: Dict[str, Dict[str, Any]] = defaultdict(lambda: {})
+
+    def __init__(self, category: str):
+        """
+        Generic GlobalTrainingStatus writer. A category is the broadest type of storage (would
+        correspond the run name and trainer name, e.g. 3DBalltest_3DBall. A key is the
+        type of status needed to be saved (e.g. Lesson Number). Finally the Value is the float value
+        attached to this stat.
+        """
+        self.category: str = category
```

Comment:
This still feels awkward having a mix of instance and static data. Why not drop `self.category` and add an extra argument to `restore_parameter_state` and `store_parameter_state`?

Reply:
I went this way to keep it inline with the StatsReporter, and so that if a trainer uses it multiple times it doesn't need to continue to pass in the category. 

Don't have a strong preference since it's used much less frequently than StatsReporter; I'm OK with changing it to a parameter to the class methods.

Reply:
Changed it to two args for `get_parameter_state` and `set_parameter_state`.

Reply:
Can you remove self.category (and the initializer) now?

Reply:
Yes - oversight on my part. Removed both.